Sources/SwiftDocC/Infrastructure/ContentCache.swift

```diff
@@ -0,0 +1,93 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+extension DocumentationContext {
+    /// A cache for symbol and page content.
+    struct ContentCache<Value> {
+        private var storage = [ResolvedTopicReference: Value]()
+        private var symbolIndex = [String: ResolvedTopicReference]()
+        
+        /// Accesses the value for a given reference.
+        /// - Parameter reference: The reference to find in the cache.
+        subscript(reference: ResolvedTopicReference) -> Value? {
+            // Avoid copying the values if possible
+            _read { yield storage[reference] }
+            _modify { yield &storage[reference] }
+        }
+        
+        /// Adds a value to the cache for a given reference _and_ symbol ID.
+        /// - Parameters:
+        ///   - value: The value to add to the cache.
+        ///   - reference: The reference associated with that value.
+        ///   - symbolID: The symbol ID associated with that value.
+        mutating func add(value: Value, reference: ResolvedTopicReference, symbolID: String) {
+            symbolIndex[symbolID] = reference
+            storage[reference] = value
+        }
+        
+        /// Accesses the reference for a given symbol ID.
+        /// - Parameter symbolID: The symbol ID to find in the cache.
+        func reference(symbolID: String) -> ResolvedTopicReference? {
+            symbolIndex[symbolID]
+        }
+        
+        /// Accesses the value for a given symbol ID.
+        /// - Parameter symbolID: The symbol ID to find in the cache.
+        subscript(symbolID: String) -> Value? {
+            // Avoid copying the values if possible
+            _read { yield symbolIndex[symbolID].map { storage[$0]! } }
+        }
+        
+        /// Reserves enough space to store the specified number of values.
+        mutating func reserveCapacity(_ minimumCapacity: Int, reserveSymbolIDCapacity: Bool) {
+            storage.reserveCapacity(minimumCapacity)
+            if reserveSymbolIDCapacity {
+                symbolIndex.reserveCapacity(minimumCapacity)
+            }
+        }
+        
+        /// Returns a list of all the references in the cache.
+        var references: [ResolvedTopicReference] {
+            return Array(storage.keys)
```

Comment:
I would argue that 90% of the time, the developer only wants something that they can access elements of and something that they can iterate over, it doesn't specifically need to be an `Array`. 

When we update to a newer Swift version—at least 5.7 to be specific—then this can be written as `some Collection<ResolvedTopicReference>` instead which allows the dictionary keys to be returned directly—avoiding copying the keys to create an array—without the caller needing to know that the type is `Dictionary<ResolvedTopicReference, Value>.Keys`
```swift
var references: some Collection<ResolvedTopicReference> {
    return storage.keys
}
```